flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java

```diff
@@ -0,0 +1,263 @@
+
+		bootstrap = new Bootstrap();
+		bootstrap
+			.group(group)
+			.channel(NioSocketChannel.class)
+			.handler(initializer);
+
+		LOG.info("Rest client endpoint started.");
+	}
+
+	public void shutdown() {
+		if (bootstrap != null) {
+			if (bootstrap.group() != null) {
+				bootstrap.group().shutdownGracefully();
+			}
+		}
+	}
+
+	public <M extends MessageHeaders<R, P, U>, U extends ParameterMapper, R extends RequestBody, P extends ResponseBody> CompletableFuture<P> sendRequest(M messageHeaders, U urlResolver, R request) throws IOException {
+		Preconditions.checkNotNull(messageHeaders);
+		Preconditions.checkNotNull(request);
+
+		String targetUrl = ParameterMapper.resolveUrl(
+			messageHeaders.getTargetRestEndpointURL(),
+			urlResolver.mapPathParameters(messageHeaders.getPathParameters()),
+			urlResolver.mapQueryParameters(messageHeaders.getQueryParameters())
+		);
+
+		LOG.debug("Sending request of class {} to {}", request.getClass(), targetUrl);
+		// serialize payload
+		StringWriter sw = new StringWriter();
+		objectMapper.writeValue(sw, request);
+		ByteBuf payload = Unpooled.wrappedBuffer(sw.toString().getBytes(ConfigConstants.DEFAULT_CHARSET));
+
+		// create request and set headers
+		FullHttpRequest httpRequest = new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, messageHeaders.getHttpMethod().getNettyHttpMethod(), targetUrl, payload);
+		httpRequest.headers()
+			.add(HttpHeaders.Names.CONTENT_LENGTH, payload.capacity())
+			.add(HttpHeaders.Names.CONTENT_TYPE, "application/json; charset=" + ConfigConstants.DEFAULT_CHARSET.name())
+			.set(HttpHeaders.Names.HOST, configuredTargetAddress + ":" + configuredTargetPort)
+			.set(HttpHeaders.Names.CONNECTION, HttpHeaders.Values.CLOSE);
+
+		synchronized (this) {
+			// This ensures strict sequential processing of requests.
+			// If we send new requests immediately we can no longer make assumptions about the order in which responses
+			// arrive, due to which the handler cannot know which future he should complete (not to mention what response
+			// type to read).
+			CompletableFuture<P> nextFuture = lastFuture
+				.handleAsync((f, e) -> submitRequest(httpRequest, messageHeaders), directExecutor)
+				.thenCompose((future) -> future);
+
+			lastFuture = nextFuture;
+			return nextFuture;
+		}
+	}
+
+	private <M extends MessageHeaders<R, P, U>, U extends ParameterMapper, R extends RequestBody, P extends ResponseBody> CompletableFuture<P> submitRequest(FullHttpRequest httpRequest, M messageHeaders) {
+		CompletableFuture<P> responseFuture = handler.expectResponse(messageHeaders.getResponseClass());
+
+		try {
+			// write request
+			Channel channel = bootstrap.connect(configuredTargetAddress, configuredTargetPort).sync().channel();
+			channel.writeAndFlush(httpRequest);
+			channel.closeFuture();
+		} catch (InterruptedException e) {
+			return FutureUtils.completedExceptionally(e);
+		}
+		return responseFuture;
+	}
+
+	@ChannelHandler.Sharable
```

Comment:
Let's not make it sharable and instead every handler has its own future it completes.